boxtree/tree_build.py

```diff
@@ -46,6 +56,240 @@ class MaxLevelsExceeded(RuntimeError):
+
+
+def vec_of_signs(dim, i):
+    """The sign vector is obtained by converting i to a dim-bit binary.
+    """
+    # e.g. bin(10) = '0b1010'
+    binary_digits = [int(bd) for bd in bin(i)[2:]]
+    n = len(binary_digits)
+    assert n <= dim
+    return np.array([0]*(dim-n) + binary_digits) * 2 - 1
+
+
+# {{{ refine/coarsen a tree of boxes
+
+def refined(tob, refine_flags):
+    """Make a refined copy of `tob` where boxes flagged with `refine_flags` are
+    refined.
+    """
+    return refined_and_coarsened(tob, refine_flags, None)
+
+
+def uniformly_refined(tob):
+    """Make a uniformly refined copy of `tob`.
+    """
+    refine_flags = np.zeros(tob.nboxes, bool)
+    refine_flags[tob.get_leaf_flags()] = 1
+    return refined(tob, refine_flags)
+
+
+def coarsened(tob, coarsen_flags):
+    """Make a coarsened copy of `tob` where boxes flagged with `coarsen_flags`
+    are coarsened.
+    """
+    return refined_and_coarsened(tob, None, coarsen_flags)
+
+
+def refined_and_coarsened(tob, refine_flags=None, coarsen_flags=None):
+    """Make a refined/coarsened copy. When children of the same parent box
+    are marked differently, the refinement flag takes priority.
+
+    Both refinement and coarsening flags can only be set of leaves.
+    To prevent drastic mesh change, coarsening is only executed when a leaf
+    box is marked for coarsening, and its parent's children are all leaf
+    boxes (so that change in the number of boxes is bounded per box flagged).
+
+    :arg refine_flags: a boolean array of size `nboxes`.
+    :arg coarsen_flags: a boolean array of size `nboxes`.
+    :returns: a processed copy of the tree.
+    """
+    nchildren = 2**tob.dim
+
+    if refine_flags is None:
+        refine_flags = np.zeros(tob.nboxes, bool)
+    if refine_flags[~tob.get_leaf_flags()].any():
+        raise ValueError("attempting to split non-leaf")
+
+    if coarsen_flags is None:
+        coarsen_flags = np.zeros(tob.nboxes, bool)
+    if coarsen_flags[~tob.get_leaf_flags()].any():
+        raise ValueError("attempting to coarsen non-leaf")
+
+    # ------------------- refinement always respects the flags
+
+    refine_parents, = np.where(refine_flags)
+    n_new_boxes = len(refine_parents) * nchildren
+    nboxes_new = tob.nboxes + n_new_boxes
+
+    child_box_starts = (
+            tob.nboxes
+            + nchildren * np.arange(len(refine_parents)))
+
+    refine_parents_per_child = np.empty(
+            (nchildren, len(refine_parents)),
+            np.intp)
+    refine_parents_per_child[:] = refine_parents.reshape(-1)
+    refine_parents_per_child = refine_parents_per_child.reshape(-1)
+
+    box_parents = resized_array(tob.box_parent_ids, nboxes_new)
+    box_centers = resized_array(tob.box_centers, nboxes_new)
+    box_children = resized_array(tob.box_child_ids, nboxes_new)
+    box_levels = resized_array(tob.box_levels, nboxes_new)
+
+    # new boxes are appended at the end, so coarsen_flags are still meaningful
```

Comment:
Levels in `Tree` are contiguous, I think. So this is a fundamentally different data structure.